pallets/watchtower/src/lib.rs

```diff
@@ -0,0 +1,1195 @@
+                    consensus_key
+                );
+                return Ok(());
+            }
+
+            if root_hash == sp_core::H256::zero() {
+                log::warn!(
+                    target: "runtime::watchtower::notification",
+                    "Received empty root hash for summary: {:?}",
+                    consensus_key
+                );
+            }
+
+            VotingStartBlock::<T>::insert(
+                &consensus_key,
+                frame_system::Pallet::<T>::block_number(),
+            );
+            PendingValidationRootHash::<T>::insert(&consensus_key, root_hash);
+
+            log::info!(
+                target: "runtime::watchtower::notification",
+                "Started voting period for summary: instance={:?}, root_id={:?}, hash={:?}",
+                instance, root_id, root_hash
+            );
+
+            Ok(())
+        }
+
+        fn process_pending_validations() {
+            log::debug!(target: "runtime::watchtower::ocw", "Processing pending validations");
+
+            let current_block = frame_system::Pallet::<T>::block_number();
+            let voting_period = VotingPeriod::<T>::get();
+
+            for (consensus_key, start_block) in VotingStartBlock::<T>::iter() {
+                let (instance, root_id) = consensus_key.clone();
+
+                if VoteConsensusReached::<T>::get(&consensus_key) {
+                    continue;
+                }
+
+                if current_block > start_block + voting_period {
+                    log::warn!(
+                        target: "runtime::watchtower::ocw",
+                        "Voting period expired for {:?}, skipping OCW validation",
+                        consensus_key
+                    );
+                    continue;
+                }
+
+                if let Some(root_hash) = PendingValidationRootHash::<T>::get(&consensus_key) {
+                    log::info!(
+                        target: "runtime::watchtower::ocw",
+                        "Processing OCW validation for {:?}, root_hash: {:?}",
+                        consensus_key, root_hash
+                    );
+
+                    Self::perform_ocw_recalculation(instance, root_id, root_hash);
+
+                    PendingValidationRootHash::<T>::remove(&consensus_key);
+                } else {
+                    log::warn!(
+                        target: "runtime::watchtower::ocw",
+                        "No root hash found for active voting period: {:?}",
+                        consensus_key
+                    );
+                }
+            }
+        }
+
+        fn perform_ocw_recalculation(
+            instance: SummarySourceInstance,
+            root_id: WatchtowerRootId<BlockNumberFor<T>>,
+            onchain_root_hash: WatchtowerOnChainHash,
+        ) {
+            let maybe_node_info = Self::get_node_from_signing_key();
+            let (node, signing_key) = match maybe_node_info {
```

Comment:
I think we should get the node signing key when the OCW starts and if we don't get a valid signing key, we won't bother doing any other processing and avoid unnecessary work. If we manage to get the node signing key, we can pass it around the other function calls.